### PR TITLE
Allow total_estimated_cells to be None (#2271)

### PR DESCRIPTION
**Pull request by amarjandu**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [x] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

File: test/indexer/__init__.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def _load_canned_result(self, bundle_fqid: BundleFQID) -> MutableJSONs:
             hit['_index'] = config.es_index_name(catalog=self.catalog,
                                                  entity_type=index_name.entity_type,
                                                  aggregate=index_name.aggregate)
+            hit.pop('sort')
```
Comment (danielsotirhos) [inline test/indexer/__init__.py:95]: This and/or the removal of `sort` in `_get_all_hits()` could use a short comment explaining why we need to remove this field for our test.

File: test/service/test_response.py [hunk 2]
```diff
@@ -1442,7 +1442,7 @@ def test_ordering(self):
             hit_sort_values = {}
             for order, response in responses.items():
                 response.raise_for_status()
-                hit_sort_values[order] = [accessor(hit) for hit in response.json()['hits']]
+                hit_sort_values[order] = [accessor(hit) for hit in response.json()['hits'] if accessor(hit) is not None]
```
Comment (danielsotirhos) [inline test/service/test_response.py:1445]: Now that the contents of `hit_sort_values` is filtered (`if accessor(hit) is not None`), adding verification that the lists aren't empty would make the sorting verification stronger.
```
            self.assertGreater(len(hit_sort_values['asc']), 0)
            self.assertGreater(len(hit_sort_values['desc']), 0)
```

File: test/indexer/test_hca_indexer.py [hunk 3]
```diff
@@ -1025,6 +1026,31 @@ def test_accessions_fields(self):
...
+                entity_type, aggregate = self._parse_index_name(hit)
+                if entity_type == 'projects' and aggregate:
+                    self.assertEqual(expected_cell_count,
+                                     one(hit['_source']['contents']['cell_suspensions'])['total_estimated_cells'])
+                    self.assertEqual(expected_cell_count_,
+                                     one(hit['_source']['contents']['cell_suspensions'])['total_estimated_cells_'])
+                    self.assertEqual(expected_organ_parts,
+                                     one(hit['_source']['contents']['cell_suspensions'])['organ_part'])
+
+        no_cells_bundle = self._load_canned_bundle(BundleFQID('587d74b4-1075-4bbf-b96a-4d1ede0481b2',
```
Comment (danielsotirhos) [inline test/indexer/test_hca_indexer.py:1044]:
```suggestion
        # This bundle has a 'cell_suspension' yet has no 'total_estimated_cells' field.
        no_cells_bundle = self._load_canned_bundle(BundleFQID('587d74b4-1075-4bbf-b96a-4d1ede0481b2',
```

File: test/indexer/test_hca_indexer.py [hunk 4]
```diff
@@ -1025,6 +1026,31 @@ def test_accessions_fields(self):
...
+                                                              '2018-09-14T133314.453337Z'))
+        self._index_bundle(no_cells_bundle)
+        assert_cell_suspension(None, ['temporal lobe'], self._get_all_hits())
+        has_cells_bundle = self._load_canned_bundle(BundleFQID('3db604da-940e-49b1-9bcc-25699a55b295', ''))
```
Comment (danielsotirhos) [inline test/indexer/test_hca_indexer.py:1048]:
```suggestion
        # This bundle has a 'cell_suspension' with a 'total_estimated_cells' field.
        has_cells_bundle = self._load_canned_bundle(BundleFQID('3db604da-940e-49b1-9bcc-25699a55b295', ''))
```

Comment (danielsotirhos) [inline test/indexer/test_hca_indexer.py:1048]:
```suggestion
        has_cells_bundle = self._load_canned_bundle(BundleFQID('3db604da-940e-49b1-9bcc-25699a55b295', '2018-11-02T184048.983513Z'))
```

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
Looks great. Just a couple suggestions to add a few comments.
Also, is there a current canned bundle that could be modified to set `total_estimated_cells` to `0` without breaking any existing tests? If so adding that in to `test_no_cell_count_contributions()` would cover all the possible values (None, 0, >0).

File: test/indexer/test_hca_indexer.py (hunk 4 above)
Comment (amarjandu) [reply test/indexer/test_hca_indexer.py:1048]: Good catch, for some reason I was thinking that the version stamp was significant during the loading operation.

**Comment by amarjandu** [comment]
@danielsotirhos ask and maybe you shall get what you want. ¯\_(ツ)_/¯

```
grep -H '"total_estimated_cells": 0,' test/indexer/data/*
test/indexer/data/d0e17014-9a58-4763-9e66-59894efbdaa8.metadata.json:        "total_estimated_cells": 0,
test/indexer/data/d0e17014-9a58-4763-9e66-59894efbdaa8.metadata.json:        "total_estimated_cells": 0,
test/indexer/data/d0e17014-9a58-4763-9e66-59894efbdaa8.metadata.json:        "total_estimated_cells": 0,
test/indexer/data/d0e17014-9a58-4763-9e66-59894efbdaa8.metadata.json:        "total_estimated_cells": 0,
```

I'll try to add one in.

**Comment by amarjandu** [comment]
@danielsotirhos I gave it a rough shot to add it in, it looks like the organs are different (Plasma vs brain) and there are almost ~650 metadata objects within the bundle. Further I'm not sure if the `0` case need to be explicitly tested as it should be handled as any other integer. 

If you feel strongly about this I don't doing more metadata/manifest modifications to this bundle.

**Comment by danielsotirhos** [comment]
Regarding the `0` case I agree with your assessment. We might want to add a doctest to SumAccumulator at some point though to prove the `0` and `None` cases.

**Review by danielsotirhos — APPROVED** [review]
LGTM.

File: test/indexer/test_hca_indexer.py [hunk 5]
```diff
@@ -1027,6 +1028,34 @@ def test_accessions_fields(self):
...
 
+    def test_no_cell_count_contributions(self):
+        def assert_cell_suspension(expected_cell_count, expected_organ_parts, hits):
+            expected_cell_count_ = expected_cell_count
+            if expected_cell_count is None:
+                expected_cell_count = null_int.null_int
+            for hit in hits:
+                entity_type, aggregate = self._parse_index_name(hit)
+                if entity_type == 'projects' and aggregate:
+                    self.assertEqual(expected_cell_count,
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1039]: test would pass if there were no aggregate project hits

and I don't understand why you incrementally indexing the bundles without clearing the index. Each time this functions is called, the previous bundles' hits are asserted, too. How can this even work?

Comment (hannes-ucsc) [reply test/indexer/test_hca_indexer.py:1039]: consider a different approach, one that fingerprints the hits and asserts the fingerprint. I can explain in parking lot.

File: test/indexer/test_hca_indexer.py [hunk 6]
```diff
@@ -1027,6 +1028,34 @@ def test_accessions_fields(self):
...
+        # This bundle has a 'cell_suspension' yet has no 'total_estimated_cells' field.
+        no_cells_bundle = self._load_canned_bundle(BundleFQID('587d74b4-1075-4bbf-b96a-4d1ede0481b2',
+                                                              '2018-09-14T133314.453337Z'))
+        self._index_bundle(no_cells_bundle)
+        assert_cell_suspension(None, ['temporal lobe'], self._get_all_hits())
+        # This bundle has a 'cell_suspension' with a 'total_estimated_cells' field.
+        # The version has been chosen to ensure that the project is correctly aggregated.
+        has_cells_bundle = self._load_canned_bundle(BundleFQID('3db604da-940e-49b1-9bcc-25699a55b295',
+                                                               '2018-11-02T184048.983513Z'))
+        has_cells_bundle.metadata_files['project_0.json'] = no_cells_bundle.metadata_files['project_0.json']
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1055]: This needs an explanation.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by hannes-ucsc** [comment]
Also would like to see split test-then-fix commits pushed separately.

**Comment by amarjandu** [comment]
> Also would like to see split test-then-fix commits pushed separately.

It was not my intention to make this a `test-then-fix` commit, although I can break it down to do that.

File: test/indexer/__init__.py [hunk 7]
```diff
@@ -92,6 +92,10 @@ def _load_canned_result(self, bundle_fqid: BundleFQID) -> MutableJSONs:
             hit['_index'] = config.es_index_name(catalog=self.catalog,
                                                  entity_type=index_name.entity_type,
                                                  aggregate=index_name.aggregate)
+            # The documents` sort position within the index can be different
```
Comment (hannes-ucsc) [inline test/indexer/__init__.py:95]: Could be better. Discuss on parking lot.

File: test/indexer/test_hca_indexer.py [hunk 8]
```diff
@@ -1027,6 +1031,41 @@ def test_accessions_fields(self):
             self.assertEqual(['E-AAAA-00'], project['array_express_accessions'])
             self.assertEqual(['PRJNA000000'], project['insdc_study_accessions'])
 
+    def test_no_cell_count_contributions(self):
+        def assert_cell_suspension(expected_result: Dict, hits):
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1035]: Dict is generic

File: test/indexer/test_hca_indexer.py [hunk 9]
```diff
@@ -1027,6 +1031,41 @@ def test_accessions_fields(self):
...
+            fields = {'total_estimated_cells', 'total_estimated_cells_', 'organ_part'}
+            for hit in hits:
+                entity_type, aggregate = self._parse_index_name(hit)
+                if entity_type == 'projects' and aggregate:
+                    cell_suspension = one(hit['_source']['contents']['cell_suspensions'])
+            self.assertEqual(expected_result, {field: cell_suspension[field] for field in fields})
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1041]: too complicated

File: test/service/test_manifest.py [hunk 10]
```diff
@@ -375,7 +375,7 @@ def test_terra_bdbag_manifest(self):
                 'bundle_version': '2018-09-14T133314.453337Z',
                 'cell_suspension__provenance__document_id': '377f2f5a-4a45-4c62-8fb0-db9ef33f5cf0',
                 'cell_suspension__biomaterial_core__biomaterial_id': 'Q4_DEMO-cellsus_SAMN02797092',
-                'cell_suspension__estimated_cell_count': '0',
+                'cell_suspension__estimated_cell_count': 'None',
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:378]:
```suggestion
                'cell_suspension__estimated_cell_count': '',
```

File: test/service/test_response.py [hunk 11]
```diff
@@ -1442,7 +1442,8 @@ def test_ordering(self):
...
```
Comment (hannes-ucsc) [inline test/service/test_response.py:1445]: line too long

Comment (hannes-ucsc) [reply test/service/test_response.py:1445]: use filter()

Comment (hannes-ucsc) [reply test/service/test_response.py:1445]: and map()

Comment (hannes-ucsc) [reply test/service/test_response.py:1445]: Most importantly, why are we excluding None values from the asserted sort order? The sorting should be total and cover None values.

File: test/indexer/test_hca_indexer.py [hunk 12]
```diff
@@ -1027,6 +1031,41 @@ def test_accessions_fields(self):
...
+        no_cells_bundle = self._load_canned_bundle(BundleFQID('587d74b4-1075-4bbf-b96a-4d1ede0481b2',
+                                                              '2018-09-14T133314.453337Z'))
+        self._index_bundle(no_cells_bundle)
+        assert_cell_suspension({
+            'total_estimated_cells': null_int.null_int,
+            'total_estimated_cells_': None,
+            'organ_part': ['temporal lobe']
+        },
+            self._get_all_hits())
+        # This bundle has a 'cell_suspension' with a 'total_estimated_cells'
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1053]: Could be better.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_hca_indexer.py (hunk 8 above)
Comment (amarjandu) [reply test/indexer/test_hca_indexer.py:1035]: opted to omit, method is only used within this testcase.

File: test/service/test_response.py (hunk 11 above)
Comment (amarjandu) [reply test/service/test_response.py:1445]: Cant compare `Nonetype` and `int`,  might be better to replace `None` here with `float('-inf')` so we can keep ensure the sort comparisons.

File: test/service/test_manifest.py (hunk 10 above)
Comment (amarjandu) [reply test/service/test_manifest.py:378]: needs discussion

File: test/service/test_response.py (hunk 11 above)
Comment (amarjandu) [reply test/service/test_response.py]: used none_safe_key

File: test/indexer/test_hca_indexer.py [hunk 13]
```diff
@@ -1047,6 +1047,41 @@ def test_accessions_fields(self):
...
+        # field. The bundles are incrementally index to observe that the
+        # estimated_cell_count can be set from None to a value.
+        has_cells_bundle = self._load_canned_bundle(BundleFQID('3db604da-940e-49b1-9bcc-25699a55b295',
+                                                               '2018-09-14T133314.453337Z'))
+        # Project metadata is changed here to ensure that the project_aggregate
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1074]: line length?
```suggestion
        # Project metadata is changed here to ensure that the project aggregate
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Many lines in this PR look longer than 80 stripped.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I'll push fixes.

I hope I didn't break anything. If I did, and for simplicity's sake, you can squash my commits into yours and add any fixes.

I'd like to see the test and fix pushed separately one more time, please.

File: test/indexer/test_hca_indexer.py [hunk 14]
```diff
@@ -1047,6 +1047,50 @@ def test_accessions_fields(self):
...
+        target_metadata = has_cells_bundle.metadata_files
+        source_metadata = no_cells_bundle.metadata_files
+        target_metadata['project_0.json'] = source_metadata['project_0.json']
+        self._index_bundle(has_cells_bundle)
+        assert_cell_suspension({
+            'total_estimated_cells': 10000,
+            'total_estimated_cells_': 10000,
+            'organ_part': ['amygdala', 'temporal lobe']
+        },
+            self._get_all_hits())
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply test/indexer/test_hca_indexer.py:1092]: opted to extract the first argument into a variable.

**Comment by amarjandu** [comment]
@hannes-ucsc I don't see your commits pushed up to this branch, the commit history still show that I have authored all 3 commits, I don't see a notification within the thread for commits that were pushed up by you.

I have fixup! commits in place for both the tests and the fix, but I'll hold off on pushing changes.

**Comment by hannes-ucsc** [comment]
Sorry, I forgot to push. I did just now.

**Comment by amarjandu** [comment]
Screen Shot 2021-01-29 at 10 13 41 AM

the ordering of the commits and the distinction of the tests/fixes seem to be out of order, I'm going organize/squash the commits then look for any needed fixes.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Where are my contributions?

**Comment by amarjandu** [comment]
> Where are my contributions?

Whoops, I had to resurrect them from the ether. 🧙🏽‍♂️ 

It looks like they were lost when I attempted to squash them down. I also tried to use the [Co-authored feature ](https://docs.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors, but using this looses the blame tracking.

File: test/indexer/test_hca_indexer.py [hunk 15]
```diff
@@ -1047,6 +1047,58 @@ def test_accessions_fields(self):
...
+                              for hit in hits
+                              if ('projects', True) == self._parse_index_name(hit))
+            cell_suspension = one(project_hit['_source']['contents']['cell_suspensions'])
+            actual_result = {
+                field: cell_suspension[field]
+                for field in {
+                    'total_estimated_cells',
+                    'total_estimated_cells_',
+                    'organ_part'
+                }
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1062]:
```suggestion
                for field in expected.keys()
```

**Review by hannes-ucsc — COMMENTED** [review]
One teeny nit. No bump.

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — COMMENTED** [review]
Since this potentially affects aggregate documents, it should require reindexing, correct?

**Comment by amarjandu** [comment]
@hannes-ucsc I've added the reindex commit tags  + labels.

**Comment by hannes-ucsc** [comment]
For demo, checkout the commit that adds the test, show that the new test fails, show that `test_terra_bdbag_manifest` also fails. Discuss the failed assertions. Checkout the merge commit and run both tests again.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**